cmd/snap-repair/cmd_done_retry_skip_test.go

```diff
@@ -60,6 +60,9 @@ func (r *repairSuite) TestStatusHappy(c *C) {
 	for _, s := range []string{"done", "skip", "retry"} {
 		rp, wp, err := os.Pipe()
 		c.Assert(err, IsNil)
+		defer rp.Close()
+		defer wp.Close()
```

Comment:
I would add a comment before the 2nd defer // this will fail with bad descriptor but at a controlled time

another approach is to close wp only if the test didn't succeed, another one is to use syscall.Dup in which case worst case we leak 3 descriptors

Reply:
Why would it fail? It seems fine either way (I often re-close FDs on defer) but in this particular case I can't even see why this would be an issue here. The fd doesn't seem closed anywhere else.

Reply:
it's close by the code of the commend that does NewFile  starting from the int fd

Reply:
double Close are not a problem if the descriptor don't get out of control of the runtime, because then the runtime knows it was already closed, but the recreating a runtime File from the fd int confuses this, that's why had a bug in the first place....

1) it was closed once by the tested command code
2) a 2nd time through the finalizer attached to wp

Reply:
given the pain this was to track I'm slightly partial to the Dup solution

Reply:
Sorry, I must be missing something obvious here. I don't see any dup or any NewFile here. Also, as far as I know NewFlie doesn't dup either, but that's not even at play here as far as I can see.

Reply:
you are missing something obvious, the code that this is testing is doing the NewFile:

https://github.com/snapcore/snapd/blob/master/cmd/snap-repair/cmd_done_retry_skip.go#L58

and the defer Closing it.

I'm proposing to use a Dup in the test, there's not Dup yet anywhere.

Reply:
to be clear, the issue here is far from "obvious" though

Reply:
Aha, got it, thanks.